### PR TITLE
Include GitLab subgroups

### LINKED ISSUE #6068: Cannot fetch branches, trigger prebuilds, and configure projects in a subgroup for GitLab
### Bug description

This prevents users from a) loading branches for a project, b) triggering prebuilds from the dashboard, and c) configuring the project from the dashboard.

🅰️  No error or feedback is returned when trying to load projects branches.

🅱️  The following error is returned in the browser console when trying to trigger a prebuild from the prebuilds page or the configuration page:

```js
Error: Could not find a default branch in repository
```

| Branches | Configuration |
|-|-|
| Screenshot 2021-10-06 at 3 51 48 PM | Screenshot 2021-10-06 at 3 51 09 PM |

### Steps to reproduce

Create a project in a subgroup at GitLab.com and try adding this as a project in Gitpod.io.

### Expected behavior

_No response_

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by laushinka**
Fixes #6068

## Description
This PR allows GitLab projects in subgroups and nested subgroups to be accessible. This PR also includes additional fixes related to the [GitLab slug](https://github.com/gitpod-io/gitpod/pull/6376) issue.

## Related Issue(s)

Fixes #6068

## How to test
1. Create a subgroup and a subgroup inside that subgroup (however nested you like) in GitLab
2. Create a project within each of those subgroups
3. Add those projects in Gitpod. Run prebuilds for each of them.

## Release Notes

```release-note
Prebuilds can run for GitLab subgroup projects. 
```

### CODE REVIEW

**Comment by roboquat** [comment]
@laushinka: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by svenefftinge** [comment]
Multiple subgroups can be nested ([docs](https://docs.gitlab.com/ee/user/group/subgroups/)).
Would be great to add unit tests for the changed function.

**Comment by laushinka** [comment]
> Multiple subgroups can be nested ([docs](https://docs.gitlab.com/ee/user/group/subgroups/)). Would be great to add unit tests for the changed function.

@svenefftinge Thanks for this!

**Comment by laushinka** [comment]
/unhold

Ready for another review.

File: components/server/src/repohost/repo-url.spec.ts
Comment (gtsiolis) [inline components/server/src/repohost/repo-url.spec.ts:53]: **praise**: Thanks for adding these tests here, @laushinka! 🌟

File: components/server/src/repohost/repo-url.spec.ts
Comment (gtsiolis) [inline components/server/src/repohost/repo-url.spec.ts:55]: **issue(non-blocking)**: Cross-posting from https://github.com/gitpod-io/gitpod/pull/5362#discussion_r695079714 to bring this up again, although sounds ok to leave this out of the scope of this PR, in case we'd like to make plan any action:

> Shall we move this group under [`gitpod-io`](https://gitlab.com/gitpod-io) group?

Cc @corneliusludmann @geropl

**Review by gtsiolis — APPROVED** [review]
UX works as expected, @laushinka! 🔮 

Approving to unblock merging but holding in case :a: because of a minor non-blocking issue and :b: we need someone to take a closer look at the code.

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 9e8c123b4e2aaad7aaa0a6aad16178f93df67796</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/repohost/repo-url.spec.ts
Comment (laushinka) [reply components/server/src/repohost/repo-url.spec.ts:53]: Thanks to @svenefftinge for the reminder to write more tests :)

**Comment by JanKoehnlein** [comment]
Doesn't work for me. I created group0 and group1 inside and projects group0/bar and group0/group1/Foo. The latter can be prebuilt, the former can't.  Could be related to the fact that GitLab changed the URL of group0 to g2506 to avoid collision.

File: components/server/src/repohost/repo-url.spec.ts
Comment (laushinka) [reply components/server/src/repohost/repo-url.spec.ts:55]: I'd love to answer but I'm not sure I know what this question or this context means, even after reading the issue. Maybe something I can ask about in the next team call? Also cc-ing @JanKoehnlein in case he has context.

**Comment by laushinka** [comment]
> Could be related to the fact that GitLab changed the URL of group0 to g2506 to avoid collision.

😵 @JanKoehnlein show in a call?

**Comment by gtsiolis** [comment]
**question**: @JanKoehnlein did use the same project + commit for the two projects? Cc @laushinka 

**issue(non-blocking)**: I could not prebuild the second project when using the same commit, but this is a different issue related to prebuilds and out of the scope of this PR, right?

Got the following error:
> **Could not run prebuild.**
> A prebuild is already running for this commit.

**Comment by JanKoehnlein** [comment]
We just found out that the issues are likely connected to the fact that my project was called `Foo` and compared against the slug `foo` was not matched.

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by laushinka** [comment]
> We just found out that the issues are likely connected to the fact that my project was called `Foo` and compared against the slug `foo` was not matched.

@JanKoehnlein Just pushed. Please help test again with your test cases 🙏🏽

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [gtsiolis]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **mrsimonemms** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @mrsimonemms` in a comment when ready.

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [gtsiolis]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by JanKoehnlein** [comment]
/lgtm

Should we also remove `ProjectDB.findProjectByTeamAndName()`? In its implementation we also filter projects by name ignoring the slug, and it doesn't seem to be called anywhere.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 204efbd1b021ee8df403901e61793a4449d714f6</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6546#issuecomment-961742847" title="LGTM">JanKoehnlein</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [JanKoehnlein,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by laushinka** [comment]
> /lgtm

Did the prebuilds work now for your projects? I had issues still with a few of them, but not sure if it's because of the specific repos/configs as Chris mentioned [in this conversation](https://gitpod.slack.com/archives/C020VCB0U5A/p1636034789124100).

> Should we also remove `ProjectDB.findProjectByTeamAndName()`? In its implementation we also filter projects by name ignoring the slug, and it doesn't seem to be called anywhere.

Good point.

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6546#issuecomment-961742847" title="LGTM">JanKoehnlein</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [JanKoehnlein,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [JanKoehnlein,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 9e4daa0e7bb0ba5d55de6466f05fab50e41b1a05</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6546#pullrequestreview-797651289" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6546#issuecomment-961840762" title="LGTM">JanKoehnlein</a>*

Associated issue: *#6068*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [JanKoehnlein,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [JanKoehnlein,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by gtsiolis** [comment]
5t2ktu

**Pull request merged**
